network/src/application/db.rs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+use crate::{
+    application::types::{PeerError, PeerInfo},
+    protocols::wire::handshake::v1::SupportedProtocols,
+};
+use diem_infallible::RwLock;
+use diem_types::PeerId;
+use std::collections::{
+    hash_map::{
+        Entry,
+        Entry::{Occupied, Vacant},
+    },
+    HashMap,
+};
+
+/// Structure for storing PeerInfo
+/// HashMap based storage
+pub struct PeerDb {
+    peers: RwLock<HashMap<PeerId, PeerInfo>>,
+}
+
+impl PeerDb {
+    pub fn new() -> PeerDb {
+        PeerDb {
+            peers: RwLock::new(HashMap::new()),
+        }
+    }
+
+    /// Read a copy of the stored `PeerInfo`
+    pub fn read(&self, peer_id: &PeerId) -> Option<PeerInfo> {
+        self.peers.read().get(peer_id).cloned()
+    }
+
+    /// Read a copy of the entire state
+    pub fn read_all(&self) -> HashMap<PeerId, PeerInfo> {
+        self.peers.read().clone()
+    }
+
+    /// Insert new entry
+    pub fn insert(&self, peer_id: PeerId, protocols: SupportedProtocols) -> Result<(), PeerError> {
+        let mut peers = self.peers.write();
+        if let Vacant(e) = peers.entry(peer_id) {
+            e.insert(PeerInfo::new(protocols));
+            Ok(())
+        } else {
+            Err(PeerError::AlreadyExists)
+        }
+    }
+
+    /// Remove old entries
+    pub fn remove(&self, peer_id: PeerId) -> Result<(), PeerError> {
+        let mut peers = self.peers.write();
+        if let Occupied(e) = peers.entry(peer_id) {
+            e.remove();
+            Ok(())
+        } else {
+            Err(PeerError::NotFound)
+        }
+    }
+
+    /// Take in a function to modify the data, must handle concurrency control with the input function
+    pub fn write<F: FnOnce(&mut Entry<PeerId, PeerInfo>) -> Result<(), PeerError>>(
```

Comment:
Interesting.. My remaining concern here is how we're avoiding concurrency issues with hashmap entries? Or are we planning on not supporting concurrency? e.g., if two threads (or tasks), call this function in unison, requesting the same entry, they can both have access to it concurrently because the write lock is dropped before the entry is passed to `modifier`. I'm guessing this is desired behaviour? If so, why? Why not just grab the write lock and hold it until `modifier` completes? 

Note: I'm probably missing something :D